nighthawk/test/sequencer_test.cc

```diff
@@ -0,0 +1,228 @@
+        dispatcher_(api_.allocateDispatcher()), callback_test_count_(0), frequency_(10_Hz),
+        interval_(std::chrono::duration_cast<std::chrono::milliseconds>(frequency_.interval())),
+        test_number_of_intervals_(5), sequencer_target_(std::bind(&SequencerTestBase::callback_test,
+                                                                  this, std::placeholders::_1)),
+        clock_updates_(0) {
+    platform_util_.setTimeSystem(this->time_system_);
+    // When yieldCurrentThread() is called we are in a tight spin loop.
+    // SimulatedTimeAwarePlatformUtil moves the simulated time forward, avoiding the tests hanging.
+    ON_CALL(platform_util_, yieldCurrentThread())
+        .WillByDefault(testing::Invoke(&platform_util_, &MockPlatformUtil::yieldWithSimulatedTime));
+  }
+
+  virtual ~SequencerTestBase() = default;
+
+  void moveClockForwardOneInterval() {
+    time_system_.setMonotonicTime(time_system_.monotonicTime() + interval_);
+    clock_updates_++;
+  }
+
+  bool callback_test(std::function<void()> f) {
+    callback_test_count_++;
+    f();
+    return true;
+  }
+  bool timeout_test(std::function<void()> /* f */) {
+    callback_test_count_++;
+    // We don't call f(); which will cause the sequencer to think there is in-flight work.
+    return true;
+  }
+  bool saturated_test(std::function<void()> /* f */) { return false; }
+
+  MockPlatformUtil platform_util_;
+  Envoy::Stats::IsolatedStoreImpl store_;
+  Envoy::Event::SimulatedTimeSystem time_system_;
+  Envoy::Api::Impl api_;
+  // TODO(oschaaf): test with MockDispatcher?
+  Envoy::Event::DispatcherPtr dispatcher_;
+  int callback_test_count_;
+  const Frequency frequency_;
+  const std::chrono::milliseconds interval_;
+  const uint64_t test_number_of_intervals_;
+  std::unique_ptr<RateLimiter> rate_limiter_;
+  SequencerTarget sequencer_target_;
+  uint64_t clock_updates_;
+};
+
+class SequencerTest : public SequencerTestBase {
+public:
+  SequencerTest() { rate_limiter_ = std::make_unique<MockRateLimiter>(); }
+  MockRateLimiter& getRateLimiter() const { return dynamic_cast<MockRateLimiter&>(*rate_limiter_); }
+};
+
+TEST_F(SequencerTest, EmptyCallbackAsserts) {
+  SequencerTarget callback_empty;
+
+  ASSERT_DEATH(SequencerImpl sequencer(platform_util_, *dispatcher_, time_system_, getRateLimiter(),
+                                       callback_empty, 1s, 1s),
+               "No SequencerTarget");
+}
+
+TEST_F(SequencerTest, StartingTwiceAsserts) {
+  SequencerImpl sequencer(platform_util_, *dispatcher_, time_system_, getRateLimiter(),
+                          sequencer_target_, 1s, 1s);
+  EXPECT_CALL(getRateLimiter(), tryAcquireOne());
+  sequencer.start();
+  ASSERT_DEATH(sequencer.start(), "");
+}
+
+TEST_F(SequencerTest, WaitButNotStartedAsserts) {
+  SequencerImpl sequencer(platform_util_, *dispatcher_, time_system_, getRateLimiter(),
+                          sequencer_target_, 1s, 1s);
+  ASSERT_DEATH(sequencer.waitForCompletion(), "");
```

Comment:
Validating `ASSERT` behavior is a bit strange, since you're driving the system into behavior that is explicitly never expected to happen. I think you can safely skip these.